File: arrays_hashing/product_of_array_except_self/js/product_of_array_except_self.js
const nums = [ 1, 2, 3, 4 ];

const productExceptSelf = (nums) => {
  answer = new Array(nums.length).fill(0);

  let prev = 1;
  for(let i in nums) {
    answer[i] = prev;
    prev *= nums[i];
  }

  let next = 1;
  for(let i = nums.length-1; i >= 0; i--) {
    answer[i] *= next;
    next *= nums[i];
  }
  
  return answer;
};

console.log(productExceptSelf(nums));

File: arrays_hashing/product_of_array_except_self/cpp/product_of_array_except_self.cpp
#include<iostream>
#include<vector>

using std::cout;
using std::vector;

// Time O(n)
// Space O(1)

class Solution {
public:
    static vector<int> productExceptSelf(const vector<int>& nums) {
        vector<int> answer(nums.size(), 1);

        // multiply answer[i] by the product of all previous elements
        int prev = 1;
        for(int i = 0; i < nums.size(); i++) {
            answer[i] = prev;
            prev *= nums[i];
        }

        // multiply answer[i] by the product of all following elements
        int next = 1;
        for(int i = nums.size()-1; i >= 0; i--) {
            answer[i] *= next;
            next *= nums[i];
        }

        return answer;
    }
};

int main() {
    vector<int> nums = { 1, 2, 3, 4 };

    auto answer = Solution::productExceptSelf(nums);

    for(auto& n : answer) {
        cout << n << '\n';
    }
}
